src/x86/ssse3.rs

```diff
@@ -61,6 +62,54 @@ pub unsafe fn _mm_shuffle_epi8(a: u8x16, b: u8x16) -> u8x16 {
     pshufb128(a, b)
 }
 
+/// Concatenate 16-byte blocks in `a` and `b` into a 32-byte temporary result,
+/// shift the result right by `n` bytes, and return the low 16 bytes.
+#[inline(always)]
+#[target_feature = "+ssse3"]
+#[cfg_attr(test, assert_instr(palignr, n = 15))]
+pub unsafe fn _mm_alignr_epi8(a: i8x16, b: i8x16, n: i32) -> i8x16 {
+    let n = n as u32;
+    // If palignr is shifting the pair of vectors more than the size of two
+    // lanes, emit zero.
+    if n > 32 {
+        return i8x16::splat(0);
+    }
+    // If palignr is shifting the pair of input vectors more than one lane,
+    // but less than two lanes, convert to shifting in zeroes.
+    let (a, b, n) = if n > 16 {
+        (i8x16::splat(0), a, n - 16)
+    } else {
+        (a, b, n)
+    };
```

Comment:
Are you sure we need these if conditionals? Do they result in additional codegen with branches? If so, I believe that's wrong. Unfortunately we don't have the machinery to test something like that yet.

Reply:
The corresponding clang code is [here](https://github.com/llvm-mirror/clang/blob/c8b57899133eb605927d729d08ee1c12ada540d2/lib/CodeGen/CGBuiltin.cpp#L7898-L7902).

Reply:
Yeah but that's in trans operating on compile time constants, so the codegen from that is very explicit. We're in a different boat here, where we need to make sure we aren't introducing extra branches, which kind of relies heavily on the optimizer. Maybe paste the codegen for this function somewhere? (An easy way to get it is to cause the `assert_instr` test to fail.)

Reply:
When `n` is 15:
```
	 0: pushq %rbp
	 1: movq %rsp, %rbp
	 2: palignr $0xf, %xmm1, %xmm0
	 3: popq %rbp
	 4: retq
	 5: nopl (%rax)
```
When `n` is 17:
```
	 0: pushq %rbp
	 1: movq %rsp, %rbp
	 2: psrldq $0x1, %xmm0
	 3: popq %rbp
	 4: retq
	 5: nopl (%rax,%rax)
```

Reply:
Good enough for me!